File: src/python/M3L.py
#
# M3L - Magayaga Mathematical Library (v0.9.2 / July 23, 2024)
# Copyright (c) 2024 Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me)
#

import random

class M3L:
    PI = 3.141592653589793
    E = 2.718281828459045

    @staticmethod
    def positive(x):
        return x

    @staticmethod
    def negative(x):
        return -x
    
    @staticmethod
    def add(*args):
        result = 0
        for num in args:
            result += num
        return result

    @staticmethod
    def subtract(*args):
        result = args[0]
        for num in args[1:]:
            result -= num
        return result

    @staticmethod
    def multiply(*args):
        result = 1
        for num in args:
            result *= num
        return result

    @staticmethod
    def divide(*args):
        if 0 in args[1:]:
            raise ValueError("Cannot divide by zero")
        result = args[0]
        for num in args[1:]:
            result /= num
        return result
    
    @staticmethod
    def square(x):
        return x ** 2

    @staticmethod
    def cube(x):
        return x ** 3

    @staticmethod
    def sqrt(x):
        return x ** (1/2)

    @staticmethod
    def cbrt(x):
        return x ** (1/3)
    
    class trim:
        @staticmethod
        def sin(x):
            # Taylor series approximation for sin(x)
            result = 0
            for i in range(10):
                coef = (-1) ** i
                num = x ** (2 * i + 1)
                denom = M3L.factorial(2 * i + 1)
                result += coef * (num / denom)
            return result
    
        @staticmethod
        def cos(x):
            # Taylor series approximation for cos(x)
            result = 0
            for i in range(10):
                coef = (-1) ** i
                num = x ** (2 * i)
                denom = M3L.factorial(2 * i)
                result += coef * (num / denom)
            return result
    
        @staticmethod
        def tan(x):
            return M3L.sin(x) / M3L.cos(x)
    
        @staticmethod
        def csc(x):
            return 1 / M3L.sin(x)
    
        @staticmethod
        def sec(x):
            return 1 / M3L.cos(x)
    
        @staticmethod
        def cot(x):
            return 1 / M3L.tan(x)
        
        @staticmethod
        def arcsin(x, terms=10):
            if x < -1 or x > 1:
                raise ValueError("Input should be in the range [-1, 1]")
            result = x
            term = x
            x_squared = x * x
            for n in range(1, terms):
                term *= x_squared * (2 * n - 1) / (2 * n)
                result += term / (2 * n + 1)
            return result

        @staticmethod
        def arccos(x):
            if x < -1 or x > 1:
                raise ValueError("Input should be in the range [-1, 1]")
            # Abramowitz and Stegun approximation for arccos(x)
            return M3L.PI / 2 - M3L.trim.arcsin(x)
        
        @staticmethod
        def arctan(z, terms=10000):
            if z == 0:
                return 0
            step = z / terms
            total_area = 0
            for i in range(terms):
                t1 = i * step
                t2 = (i + 1) * step
                area = (1 / (1 + t1**2) + 1 / (1 + t2**2)) * step / 2
                total_area += area
            return total_area

    @staticmethod
    def factorial(n):
        if n == 0:
            return 1
        else:
            return n * M3L.factorial(n - 1)
    
    @staticmethod
    def log(base, x):
        if base <= 0 or base == 1 or x <= 0:
            raise ValueError("Invalid input for logarithm")
        return M3L.ln(x) / M3L.ln(base)

    @staticmethod
    def power(base, exp):
        return base ** exp

    @staticmethod
    def ln(x):
        if x <= 0:
            raise ValueError("Invalid input for natural logarithm")
        return M3L.integrate(lambda t: 1 / t, 1, x)
    
    class calc:
        @staticmethod
        def summation(start, end, term):
            if start > end:
                raise ValueError("Start index must be less than or equal to the end index")
            result = 0
            for i in range(start, end + 1):
                result += term(i)
            return result
        
        @staticmethod
        def product(start, end, term):
            if start > end:
                raise ValueError("Start index must be less than or equal to the end index")
            result = 1
            for i in range(start, end + 1):
                result *= term(i)
            return result
        
    @staticmethod
    def integrate(f, a, b, N=1000):
        dx = (b - a) / N
        integral = 0
        for i in range(N):
            integral += f(a + (i + 0.5) * dx) * dx
        return integral
    
    # Example code:
    #
    # You can be f(x), f(x) = M3L.sin(x).
    #
    # result = M3L.limit(f, math.pi/2, approach='right')
    # print(result)

    @staticmethod
    def limit(func, x, approach="right", epsilon=1e-6):
        if approach not in ("right", "left"):
            raise ValueError("Approach must be 'right' or 'left'")
        if approach == "right":
            x_approach = x + epsilon
        else:
            x_approach = x - epsilon
        return func(x_approach)

    @staticmethod
    def array(matrix):
        return matrix

    @staticmethod
    def linspace(start, stop, num=50):
        if num <= 0:
            raise ValueError("Number of samples must be positive")
        step = (stop - start) / (num - 1)
        return [start + i * step for i in range(num)]

    @staticmethod
    def transpose(matrix):
        return [[matrix[j][i] for j in range(len(matrix))] for i in range(len(matrix[0]))]
    
    @staticmethod
    def inv(matrix):
        n = len(matrix)
        if n != len(matrix[0]):
            raise ValueError("Matrix must be square")
        identity = [[0] * n for _ in range(n)]
        for i in range(n):
            identity[i][i] = 1
        for i in range(n):
            factor = 1 / matrix[i][i]
            for j in range(n):
                matrix[i][j] *= factor
                identity[i][j] *= factor
            for k in range(n):
                if k != i:
                    factor = matrix[k][i]
                    for j in range(n):
                        matrix[k][j] -= factor * matrix[i][j]
                        identity[k][j] -= factor * identity[i][j]
        return identity

    @staticmethod
    def dot(a, b):
        if isinstance(a[0], (list, tuple)) and isinstance(b[0], (list, tuple)):
            return [[sum(x * y for x, y in zip(row, col)) for col in zip(*b)] for row in a]
        else:
            return sum(x * y for x, y in zip(a, b))
    
    @staticmethod
    def vdot(a, b):
        if isinstance(a[0], (list, tuple)) and isinstance(b[0], (list, tuple)):
            a_flat = [elem for sublist in a for elem in sublist]
            b_flat = [elem for sublist in b for elem in sublist]
            return sum(x * y for x, y in zip(a_flat, b_flat))
        else:
            return sum(x * y for x, y in zip(a, b))

    @staticmethod
    def multi_dot(matrices):
        if len(matrices) < 2:
            raise ValueError("At least two matrices are required for multi_dot")
        result = matrices[0]
        for matrix in matrices[1:]:
            result = M3L.multiply_matrices(result, matrix)
        return result

    @staticmethod
    def random(size):
        if isinstance(size, int):
            return [random.random() for _ in range(size)]
        elif isinstance(size, tuple):
            if len(size) == 1:
                return [random.random() for _ in range(size[0])]
            elif len(size) == 2:
                return [[random.random() for _ in range(size[1])] for _ in range(size[0])]
            else:
                raise ValueError("Invalid size for random")
        else:
            raise ValueError("Invalid size for random")

    @staticmethod
    def solve(matrix, b):
        inv_matrix = M3L.inv(matrix)
        return [sum(inv_matrix[i][j] * b[j] for j in range(len(b))) for i in range(len(matrix))]
    
    @staticmethod
    def multiply_matrices(matrix1, matrix2):
        if len(matrix1[0]) != len(matrix2):
            raise ValueError("Matrix dimensions incompatible for multiplication")
        result = [[0] * len(matrix2[0]) for _ in range(len(matrix1))]
        for i in range(len(matrix1)):
            for j in range(len(matrix2[0])):
                for k in range(len(matrix2)):
                    result[i][j] += matrix1[i][k] * matrix2[k][j]
        return result
    
    @staticmethod
    def matrix_power(matrix, n):
        if n == 0:
            return [[1 if i == j else 0 for j in range(len(matrix))] for i in range(len(matrix))]
        elif n > 0:
            result = matrix
            for _ in range(n - 1):
                result = M3L.multiply_matrices(result, matrix)
            return result
        else:
            raise ValueError("Matrix power is only defined for non-negative integers")
    
    @staticmethod
    def arange(start, stop=None, step=1):
        if stop is None:
            start, stop = 0, start
        return [i for i in range(start, stop, step)]

    @staticmethod
    def reshape(matrix, new_shape):
        flat = [item for sublist in matrix for item in sublist]
        reshaped = []
        count = 0
        for i in range(new_shape[0]):
            row = []
            for j in range(new_shape[1]):
                row.append(flat[count])
                count += 1
            reshaped.append(row)
        return reshaped

    @staticmethod
    def zeros(shape):
        if isinstance(shape, int):
            return [0] * shape
        elif isinstance(shape, tuple):
            if len(shape) == 1:
                return [0] * shape[0]
            elif len(shape) == 2:
                return [[0] * shape[1] for _ in range(shape[0])]
            else:
                raise ValueError("Invalid shape for zeros")
        else:
            raise ValueError("Invalid shape for zeros")

    @staticmethod
    def tensordot(a, b, axes=2):
        if isinstance(a[0], list) and isinstance(b[0], list):
            a_flat = [elem for sublist in a for elem in sublist]
            b_flat = [elem for sublist in b for elem in sublist]
            return sum(x * y for x, y in zip(a_flat, b_flat))
        else:
            return sum(x * y for x, y in zip(a, b))

    @staticmethod
    def kronecker(matrix1, matrix2):
        result = []
        for i in range(len(matrix1)):
            row = []
            for j in range(len(matrix1[0])):
                for m in range(len(matrix2)):
                    for n in range(len(matrix2[0])):
                        row.append(matrix1[i][j] * matrix2[m][n])
            result.append(row)
        return result

    @staticmethod
    def cholesky(matrix):
        n = len(matrix)
        L = [[0] * n for _ in range(n)]
        for i in range(n):
            for j in range(i + 1):
                if i == j:
                    sum_sq = sum(L[i][k] ** 2 for k in range(j))
                    L[i][j] = M3L.sqrt(matrix[i][i] - sum_sq)
                else:
                    sum_prod = sum(L[i][k] * L[j][k] for k in range(j))
                    L[i][j] = (matrix[i][j] - sum_prod) / L[j][j]
        return L
    
    class binary:
        @staticmethod
        def bdn(binary):
            decimal = 0
            binary_str = str(binary)
            for i in range(len(binary_str)):
                bit = int(binary_str[i])
                if bit != 0 and bit != 1:
                    raise ValueError("Invalid binary number")
                decimal += bit * (2 ** (len(binary_str) - i - 1))
            return decimal
        
        @staticmethod
        def dbn(decimal):
            if decimal < 0:
                raise ValueError("Decimal number must be non-negative")
            binary = ""
            if decimal == 0:
                return "0"
            while decimal > 0:
                binary = str(decimal % 2) + binary
                decimal //= 2
            return binary
        
    @staticmethod
    def prime_factorization(number):
        factors = []
        divisor = 2
        while number > 1:
            if number % divisor == 0:
                factors.append(divisor)
                number //= divisor
            else:
                divisor += 1
        return " * ".join(str(factor) for factor in factors)
    
    class function:
        @staticmethod
        def gamma(z):
            sqrt_two_pi = M3L.sqrt(M3L.PI * 2)
            return sqrt_two_pi * ((z - 1/2) ** (z - 1/2)) / (M3L.E ** z)
        
        @staticmethod 
        def zeta(s):
            result = 0
            n = 1
            while True:
                term = 1 / (n ** s)
                if term < 1e-6:  # Adjust this threshold for desired precision
                    break
                result += term
                n += 1
                return result

File: src/lua/M3L.lua
--
-- M3L - Magayaga Mathematical Library (v0.9.2 / July 23, 2024)
-- Copyright (c) 2024-2025 Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me)
--

-- Define the M3L class
M3L = {}

-- M3L Trigonometry
M3L.Trim = {}

-- M3L Calculus
M3L.Calc = {}

-- M3L Binary
M3L.Binary = {}

-- M3L Math Function
M3L.Function = {}

-- Define class variables
M3L.PI = 3.141592653589793
M3L.TAU = 2 * M3L.PI
M3L.E = 2.718281828459045

-- Define static methods
function M3L.positive(x)
    return x
end

function M3L.negative(x)
    return -x
end

function M3L.add(...)
    local args = {...}
    local result = 0
    for _, num in ipairs(args) do
        result = result + num
    end
    return result
end

function M3L.subtract(...)
    local args = {...}
    local result = args[1]
    for i = 2, #args do
        result = result - args[i]
    end
    return result
end

function M3L.multiply(...)
    local args = {...}
    local result = 1
    for _, num in ipairs(args) do
        result = result * num
    end
    return result
end

function M3L.divide(...)
    local args = {...}
    if #args < 2 then
        error("At least two arguments are required for division")
    end
    for _, num in ipairs(args) do
        if num == 0 then
            error("Cannot divide by zero")
        end
    end
    local result = args[1]
    for i = 2, #args do
        result = result / args[i]
    end
    return result
end

function M3L.square(x)
    return x ^ 2
end

function M3L.cube(x)
    return x ^ 3
end

function M3L.sqrt(x)
    return x ^ (1/2)
end

function M3L.cbrt(x)
    return x ^ (1/3)
end

function M3L.abs(x)
    return x < 0 and -x or x
end

function M3L.gcd(a, b)
    while b ~= 0 do
        a, b = b, a % b
    end
    return M3L.abs(a)
end

function M3L.lcm(a, b)
    if a == 0 or b == 0 then
        return 0
    end
    return M3L.abs(a * b) // M3L.gcd(a, b)
end

function M3L.fma(a, b, c)
    return a * b + c
end

function M3L.floor(x)
    local i = x // 1  -- Truncate towards zero
    if x < 0 and x ~= i then
        return i - 1
    else
        return i
    end
end

function M3L.Trim.sin(x)
    -- Taylor series approximation for sin(x)
    local result = 0
    for i = 0, 9 do
        local coef = (-1) ^ i
        local num = x ^ (2 * i + 1)
        local denom = M3L.factorial(2 * i + 1)
        result = result + coef * (num / denom)
    end
    return result
end

function M3L.Trim.cos(x)
    -- Taylor series approximation for cos(x)
    local result = 0
    for i = 0, 9 do
        local coef = (-1) ^ i
        local num = x ^ (2 * i)
        local denom = M3L.factorial(2 * i)
        result = result + coef * (num / denom)
    end
    return result
end

function M3L.Trim.tan(x)
    return M3L.sin(x) / M3L.cos(x)
end

function M3L.Trim.csc(x)
    return 1 / M3L.sin(x)
end

function M3L.Trim.sec(x)
    return 1 / M3L.cos(x)
end

function M3L.Trim.cot(x)
    return 1 / M3L.tan(x)
end

function M3L.Trim.arcsin(x, terms)
    terms = terms or 10
    if x < -1 or x > 1 then
        error("Input should be in the range [-1, 1]", 2)
    end
    result = x
    term = x
    x_squared = x * x
    for n = 1, terms - 1 do
        term = term * x_squared * (2 * n - 1) / (2 * n)
        result = result + term / (2 * n + 1)
    end
    return result
end

function M3L.Trim.arccos(x)
    if x < -1 or x > 1 then
        error("Input should be in the range [-1, 1]", 2)
    end
    -- Abramowitz and Stegun approximation for arccos(x)
    return M3L.PI / 2 - M3L.Trim.arcsin(x)
end

function M3L.Trim.arctan(z, terms)
    terms = terms or 10000
    if z == 0 then
        return 0
    end
    step = z / terms
    total_area = 0
    for i = 0, terms - 1 do
        t1 = i * step
        t2 = (i + 1) * step
        area = (1 / (1 + t1^2) + 1 / (1 + t2^2)) * step / 2
        total_area = total_area + area
    end
    return total_area
end

function M3L.factorial(n)
    if n == 0 then
        return 1
    else
        return n * M3L.factorial(n - 1)
    end
end

function M3L.log(base, x)
    if base <= 0 or base == 1 or x <= 0 then
        error("Invalid input for logarithm")
    end
    return M3L.ln(x) / M3L.ln(base)
end

function M3L.power(base, exp)
    return base ^ exp
end

function M3L.ln(x)
    if x <= 0 then
        error("Invalid input for natural logarithm")
    end
    return M3L.integrate(function(t) return 1 / t end, 1, x)
end

function M3L.Calc.summation(start, finish, term)
    if start > finish then
        error("Start index must be less than or equal to the end index")
    end
    local result = 0
    for i = start, finish do
        result = result + term(i)
    end
    return result
end

function M3L.Calc.product(start, finish, term)
    if start > finish then
        error("Start index must be less than or equal to the end index")
    end
    local result = 1
    for i = start, finish do
        result = result * term(i)
    end
    return result
end

function M3L.integrate(f, a, b, N)
    N = N or 1000
    local dx = (b - a) / N
    local integral = 0
    for i = 0, N - 1 do
        integral = integral + f(a + (i + 0.5) * dx) * dx
    end
    return integral
end

function M3L.limit(func, x, approach, epsilon)
    approach = approach or "right"
    epsilon = epsilon or 1e-6
    if approach ~= "right" and approach ~= "left" then
        error("Approach must be 'right' or 'left'")
    end
    local x_approach
    if approach == "right" then
        x_approach = x + epsilon
    else
        x_approach = x - epsilon
    end
    return func(x_approach)
end

function M3L.array(matrix)
    return matrix
end

function M3L.linspace(start, stop, num)
    num = num or 50
    if num <= 0 then
        error("Number of samples must be positive")
    end
    local step = (stop - start) / (num - 1)
    local result = {}
    for i = 0, num - 1 do
        table.insert(result, start + i * step)
    end
    return result
end

function M3L.transpose(matrix)
    local result = {}
    for i = 1, #matrix[1] do
        local row = {}
        for j = 1, #matrix do
            table.insert(row, matrix[j][i])
        end
        table.insert(result, row)
    end
    return result
end

function M3L.inv(matrix)
    local n = #matrix
    if n ~= #matrix[1] then
        error("Matrix must be square")
    end
    local identity = {}
    for i = 1, n do
        identity[i] = {}
        for j = 1, n do
            identity[i][j] = 0
        end
        identity[i][i] = 1
    end
    for i = 1, n do
        local factor = 1 / matrix[i][i]
        for j = 1, n do
            matrix[i][j] = matrix[i][j] * factor
            identity[i][j] = identity[i][j] * factor
        end
        for k = 1, n do
            if k ~= i then
                factor = matrix[k][i]
                for j = 1, n do
                    matrix[k][j] = matrix[k][j] - factor * matrix[i][j]
                    identity[k][j] = identity[k][j] - factor * identity[i][j]
                end
            end
        end
    end
    return identity
end

function M3L.dot(a, b)
    if type(a[1]) == "table" and type(b[1]) == "table" then
        local result = {}
        for i = 1, #a do
            result[i] = {}
            for j = 1, #b[1] do
                local sum = 0
                for k = 1, #a[1] do
                    sum = sum + a[i][k] * b[k][j]
                end
                result[i][j] = sum
            end
        end
        return result
    else
        local sum = 0
        for i = 1, #a do
            sum = sum + a[i] * b[i]
        end
        return sum
    end
end

function M3L.vdot(a, b)
    if type(a[1]) == "table" and type(b[1]) == "table" then
        local a_flat = {}
        for _, sublist in ipairs(a) do
            for _, elem in ipairs(sublist) do
                table.insert(a_flat, elem)
            end
        end
        local b_flat = {}
        for _, sublist in ipairs(b) do
            for _, elem in ipairs(sublist) do
                table.insert(b_flat, elem)
            end
        end
        local sum = 0
        for i = 1, #a_flat do
            sum = sum + a_flat[i] * b_flat[i]
        end
        return sum
    else
        local sum = 0
        for i = 1, #a do
            sum = sum + a[i] * b[i]
        end
        return sum
    end
end

function M3L.multi_dot(matrices)
    if #matrices < 2 then
        error("At least two matrices are required for multi_dot")
    end
    local result = matrices[1]
    for i = 2, #matrices do
        result = M3L.dot(result, matrices[i])
    end
    return result
end

function M3L.random(size)
    if type(size) == "number" then
        local result = {}
        for _ = 1, size do
            table.insert(result, math.random())
        end
        return result
    elseif type(size) == "table" then
        if #size == 1 then
            local result = {}
            for _ = 1, size[1] do
                table.insert(result, math.random())
            end
            return result
        elseif #size == 2 then
            local result = {}
            for _ = 1, size[1] do
                local row = {}
                for _ = 1, size[2] do
                    table.insert(row, math.random())
                end
                table.insert(result, row)
            end
            return result
        else
            error("Invalid size for random")
        end
    else
        error("Invalid size for random")
    end
end


function M3L.solve(matrix, b)
    local inv_matrix = M3L.inv(matrix)
    local result = {}
    for i = 1, #matrix do
        local sum = 0
        for j = 1, #b do
            sum = sum + inv_matrix[i][j] * b[j]
        end
        table.insert(result, sum)
    end
    return result
end

function M3L.multiply_matrices(matrix1, matrix2)
    local result = {}
    for i = 1, #matrix1 do
        result[i] = {}
        for j = 1, #matrix2[1] do
            local sum = 0
            for k = 1, #matrix2 do
                sum = sum + matrix1[i][k] * matrix2[k][j]
            end
            result[i][j] = sum
        end
    end
    return result
end

function M3L.matrix_power(matrix, n)
    if n == 0 then
        local result = {}
        for i = 1, #matrix do
            result[i] = {}
            for j = 1, #matrix do
                result[i][j] = i == j and 1 or 0
            end
        end
        return result
    elseif n > 0 then
        local result = matrix
        for _ = 2, n do
            result = M3L.multiply_matrices(result, matrix)
        end
        return result
    else
        error("Matrix power is only defined for non-negative integers")
    end
end

function M3L.arange(start, stop, step)
    if stop == nil then
        start, stop = 0, start
    end
    step = step or 1
    local result = {}
    for i = start, stop - 1, step do
        table.insert(result, i)
    end
    return result
end

function M3L.reshape(matrix, new_shape)
    local flat = {}
    for _, sublist in ipairs(matrix) do
        for _, item in ipairs(sublist) do
            table.insert(flat, item)
        end
    end
    local reshaped = {}
    local count = 1
    for i = 1, new_shape[1] do
        local row = {}
        for j = 1, new_shape[2] do
            row[j] = flat[count]
            count = count + 1
        end
        reshaped[i] = row
    end
    return reshaped
end

function M3L.zeros(shape)
    if type(shape) == "number" then
        return {0}
    elseif type(shape) == "table" then
        if #shape == 1 then
            local result = {}
            for _ = 1, shape[1] do
                table.insert(result, 0)
            end
            return result
        elseif #shape == 2 then
            local result = {}
            for _ = 1, shape[1] do
                local row = {}
                for _ = 1, shape[2] do
                    table.insert(row, 0)
                end
                table.insert(result, row)
            end
            return result
        else
            error("Invalid shape for zeros")
        end
    else
        error("Invalid shape for zeros")
    end
end

function M3L.tensordot(a, b, axes)
    if type(a[1]) == "table" and type(b[1]) == "table" then
        local a_flat = {}
        for _, sublist in ipairs(a) do
            for _, item in ipairs(sublist) do
                table.insert(a_flat, item)
            end
        end
        local b_flat = {}
        for _, sublist in ipairs(b) do
            for _, item in ipairs(sublist) do
                table.insert(b_flat, item)
            end
        end
        local sum = 0
        for i = 1, #a_flat do
            sum = sum + a_flat[i] * b_flat[i]
        end
        return sum
    else
        local sum = 0
        for i = 1, #a do
            sum = sum + a[i] * b[i]
        end
        return sum
    end
end

function M3L.kronecker(matrix1, matrix2)
    local result = {}
    for i = 1, #matrix1 do
        for j = 1, #matrix1[1] do
            for m = 1, #matrix2 do
                for n = 1, #matrix2[1] do
                    if not result[i] then result[i] = {} end
                    if not result[i][j] then result[i][j] = {} end
                    table.insert(result[i][j], matrix1[i][j] * matrix2[m][n])
                end
            end
        end
    end
    return result
end

function M3L.cholesky(matrix)
    local n = #matrix
    local L = {}
    for i = 1, n do
        L[i] = {}
        for j = 1, n do
            L[i][j] = 0
        end
    end
    for i = 1, n do
        for j = 1, i do
            if i == j then
                local sum_sq = 0
                for k = 1, j - 1 do
                    sum_sq = sum_sq + L[i][k] ^ 2
                end
                L[i][j] = math.sqrt(matrix[i][i] - sum_sq)
            else
                local sum_prod = 0
                for k = 1, j - 1 do
                    sum_prod = sum_prod + L[i][k] * L[j][k]
                end
                L[i][j] = (matrix[i][j] - sum_prod) / L[j][j]
            end
        end
    end
    return L
end

function M3L.Binary.bdn(binary)
    local binary_str = tostring(binary)
    local decimal = 0
    for i = 1, #binary_str do
        local bit = tonumber(binary_str:sub(i, i))
        if bit ~= 0 and bit ~= 1 then
            error("Invalid binary number")
        end
        decimal = decimal + bit * (2 ^ (#binary_str - i))
    end
    return decimal
end

function M3L.Binary.dbn(decimal)
    if decimal < 0 then
        error("Decimal number must be non-negative")
    end
    if decimal == 0 then
        return "0"
    end
    local binary = ""
    while decimal > 0 do
        binary = tostring(decimal % 2) .. binary
        decimal = math.floor(decimal / 2)
    end
    return binary
end

function M3L.prime_factorization(number)
    local factors = {}
    local divisor = 2
    while number > 1 do
        if number % divisor == 0 then
            table.insert(factors, divisor)
            number = number / divisor
        else
            divisor = divisor + 1
        end
    end
    return table.concat(factors, " * ")
end

function M3L.Function.gamma(z)
    local sqrt_two_pi = math.sqrt(M3L.PI * 2)
    return sqrt_two_pi * ((z - 1/2) ^ (z - 1/2)) / (M3L.E ^ z)
end

function M3L.Function.zeta(s)
    local result = 0
    local n = 1
    while true do
        local term = 1 / (n ^ s)
        if term < 1e-6 then
            break
        end
        result = result + term
        n = n + 1
    end
    return result
end